### PR TITLE
GUA-749 activity log frontend

### PR DESCRIPTION
**Pull request by danacotoran**
### What changed
Implement the frontend for the new activity log page.

Created a new view at `/security/sign-in-history` which currently works with hard coded fixtures but will be populated with real data once the backend is completed.
The new view renders a list of signed in events, with the first event either being an account creation event, or an introductory paragraph if the user is an existing user.

Also add a section to the Security page linking to the new Sign in history page. This new section and the new page are behind a feature flag. 

[Figma design here](
https://www.figma.com/file/0JbNRTnq5ibRtfg8BMxXDL/Accounts-design-thinking?type=design&node-id=2246-25695&t=7jILGnimqvuIOgM8-0)

### Why did it change
It was supported by user research.

### Issue tracking
https://govukverify.atlassian.net/browse/GUA-749

### Environment variables or secrets
- [ ] No environment variables or secrets were added or changed

- [ ] [Application configuration](https://govukverify.atlassian.net/TODO) is up-to-date
- [ ] Documented in the README
- [ ] Added to deployment steps
- [ ] Added to local startup config

- [ ] Let the tech lead know if the Dockerfile has changed

### Other considerations
- [ ] Demo to a BA, TA, and the team.
- [x] Recorded demo shared on [#govuk-accounts-tech](https://gds.slack.com/archives/C011Y5SAY3U)
- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

### CODE REVIEW

File: deploy/template.yaml [hunk 1]
```diff
@@ -177,6 +181,7 @@ Mappings:
       SESSIONEXPIRY: "7200000"
       AMYOURACCOUNTURL: "https://www.gov.uk/account/home"
       GOVACCOUNTSPUBLISHINGAPIURL: "https://account-api.publishing.service.gov.uk"
+      SUPPORTACTIVITYLOG: "1"
```
Comment (huwd) [inline deploy/template.yaml:184]:
```suggestion
      SUPPORTACTIVITYLOG: "0"
```

File: deploy/template.yaml [hunk 2]
```diff
@@ -125,6 +125,7 @@ Mappings:
       SESSIONEXPIRY: "7200000"
       AMYOURACCOUNTURL: "https://www.staging.publishing.service.gov.uk/account/home"
       GOVACCOUNTSPUBLISHINGAPIURL: "https://account-api.staging.publishing.service.gov.uk"
+      SUPPORTACTIVITYLOG: "1"
```
Comment (huwd) [inline deploy/template.yaml:128]:
```suggestion
      SUPPORTACTIVITYLOG: "0"
```

File: src/utils/signInHistory.ts [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+
+const activityLogEventTypeLocales: Record<string, unknown> = {
+  "signed-in": "signedIn"
+}
+
+// TODO should be in a config somewhere I suppose. 
+// Should be generated using Date.now() whenever a launch date is agreed upon
+const activityLogLaunchDateInMs = 1685032269060;
+
+const data = [
```
Comment (huwd) [inline src/utils/signInHistory.ts:16]: ship an empty fixture

File: src/utils/signInHistory.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
+// import { DynamoDB } from "aws-sdk";
```
Comment (divyaparameswaran) [inline src/utils/signInHistory.ts:1]: Can we remove this commented out code please ?

File: src/utils/signInHistory.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    formattedData.push(newRow)
+  }
+
+  return formattedData;
+}
+
+export const presentSignInHistory = (
+  // subjectId: string
+): ActivityLogEvent[] => {
+  // const userServices = await getServices(subjectId);
```
Comment (divyaparameswaran) [inline src/utils/signInHistory.ts:124]: Should we remove this commented out code please .

File: src/components/sign-in-history/sign-in-history-controller.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  req: Request,
+  res: Response
+): Promise<void> {
+  const { user } = req.session;
+  const env = getAppEnv();
+  let activityData:any[] = [];
+  let showExplanation = false;
+  let data:any = [];
+  let pagination:any = {};
+  if (user && user.subjectId) {
```
Comment (divyaparameswaran) [inline src/components/sign-in-history/sign-in-history-controller.ts:16]: You can also do user?.subjectId instead.

File: src/utils/signInHistory.ts [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+const data = [
+  {
+    "event_type": "test",
+    "session_id": "asdf",
+    "user_id": "string",
+    "timestamp": "1689210000",
+    "truncated": false
+  }];
+
+export const hasExplanationParagraph = (data: Array<any>) => {
```
Comment (divyaparameswaran) [inline src/utils/signInHistory.ts:25]: There seems to be a warning from lint on this line : https://github.com/alphagov/di-account-management-frontend/actions/runs/5209313074/jobs/9398998273?pr=914#step:10:22

**Comment by divyaparameswaran** [comment]
There seems to be one "Bug" as per Sonar : https://sonarcloud.io/project/issues?resolved=false&types=BUG&pullRequest=914&id=alphagov_di-account-management-frontend&open=AYhR8yMPSms46rXhgF9k . And 2 code smells .

File: src/utils/signInHistory.ts [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+      month: "long",
+      day: "numeric",
+      year: "numeric",
+      hour: "numeric",
+      minute: "numeric",
+      hourCycle: "h12",
+      timeZone: "GB"
+    })
+
+    newRow.visitedServices = row.activities?.length && row.activities.filter((activity: any) => activity.type == "visited");;
```
Comment (divyaparameswaran) [inline src/utils/signInHistory.ts:108]: I think the type is no longer "visited" it is the type from txma event. I can make those changes when I link backend .

**Comment by danacotoran** [comment]
> There seems to be one "Bug" as per Sonar : https://sonarcloud.io/project/issues?resolved=false&types=BUG&pullRequest=914&id=alphagov_di-account-management-frontend&open=AYhR8yMPSms46rXhgF9k . And 2 code smells .

@divyaparameswaran the "smells" seem to do with the TODO comments I left, should I remove them? This is still a WIP after all.
I'm not sure how to fix the bug it's complaining about actually, it complains that the get function in the controller is an `async` one but it kind of has to be as there is an `await` as the call to dynamoDB will be asynchronous (right?). I kind of modelled this off `YourServices` seeing as it'll work in a similar way.

**Comment by divyaparameswaran** [comment]
> > There seems to be one "Bug" as per Sonar : https://sonarcloud.io/project/issues?resolved=false&types=BUG&pullRequest=914&id=alphagov_di-account-management-frontend&open=AYhR8yMPSms46rXhgF9k . And 2 code smells .
> 
> @divyaparameswaran the "smells" seem to do with the TODO comments I left, should I remove them? This is still a WIP after all. I'm not sure how to fix the bug it's complaining about actually, it complains that the get function in the controller is an `async` one but it kind of has to be as there is an `await` as the call to dynamoDB will be asynchronous (right?). I kind of modelled this off `YourServices` seeing as it'll work in a similar way.

Ah I see no worries , we should merge this but maybe I can see how to properly fix this. Ignore the TODO they will be removed in the next set of changes any way .

**Review by divyaparameswaran — APPROVED** [review]

**Pull request merged**